src/main/commandServer/settingsValidator.ts

```diff
@@ -0,0 +1,185 @@
+
+    return [needToUpdate && errors.length === 0, errors];
+  }
+
+  /**
+   * The core function for checking proposed user settings.
+   * Walks the input: the user-provided object holding the new (and existing settings) against a verifier:
+   * 1. Complains about any fields in the input that aren't in the verifier
+   * 2. Recursively walks child-objects in the input and verifier
+   * 3. Calls validation functions off the verifier
+   * @param allowedSettings - The verifier
+   * @param currentSettings - The current preferences object
+   * @param newSettings - User's proposed new settings
+   * @param errors - Builds this list up as new errors are encountered, so multiple errors can be reported.
+   * @param prefix - For error messages only, e.g. '' for root, 'kubernetes.options', etc.
+   * @returns boolean - true if there are changes that need to be applied.
+   */
+  protected checkProposedSettings(
+    allowedSettings: Record<string, validationFunc|any>,
+    currentSettings: settingsLike,
+    newSettings: settingsLike,
+    errors: string[],
+    prefix: string): boolean {
+    // Note the "busy-evaluation" form below is used to call functions for the side-effect of error-detection:
+    // changeNeeded = f(...) || changeNeeded
+    let changeNeeded = false;
+
+    for (const k in newSettings) {
+      const fqname = prefix ? `${ prefix }.${ k }` : k;
+
+      if (!(k in allowedSettings)) {
+        errors.push(`Setting name ${ fqname } isn't recognized.`);
+      } else if (typeof (allowedSettings[k]) === 'object') {
+        if (typeof (newSettings[k]) === 'object') {
+          changeNeeded = this.checkProposedSettings(allowedSettings[k], currentSettings[k], newSettings[k], errors, fqname) || changeNeeded;
+        } else {
+          errors.push(`Setting ${ fqname } should wrap an inner object, but got <${ newSettings[k] }>.`);
+        }
+      } else if (typeof (newSettings[k]) === 'object') {
+        if (allowedSettings[k] === this.checkObjectUnchanged) {
+          // Special case for things like `.WSLIntegrations` which have unknown fields.
+          changeNeeded = this.checkObjectUnchanged(newSettings[k], currentSettings[k], errors, fqname) || changeNeeded;
+        } else {
+          // newSettings[k] should be valid JSON because it came from `JSON.parse(incoming-payload)`.
+          // It's an internal error (HTTP Status 500) if it isn't.
+          errors.push(`Setting ${ fqname } should be a simple value, but got <${ JSON.stringify(newSettings[k]) }>.`);
+        }
+      } else {
+        // Throw an exception if this field isn't a function, because in the verifier all values should be
+        // either child objects or functions.
+        changeNeeded = allowedSettings[k].call(this, currentSettings[k], newSettings[k], errors, fqname) || changeNeeded;
+        if (changeNeeded && fqname === 'kubernetes.containerEngine' && newSettings[k] === 'docker') {
+          newSettings[k] = 'moby';
+        }
```

Comment:
I guess this fixup can't be implemented in `checkContainerEngine` (because it' can't modify the new settings)?

Reply:
Nope. And guess what -- same problem with versions and enabled, so back after another change....